### PR TITLE
[16.0][IMP] EPC guidance 2.1 compliant

### PR DESCRIPTION
**Pull request by mihien**
Change the structure of the address block in the xml to be compliant with the newest SEPA guidelines (as of dec 2025).

This is heavily based on #1174 , keeping just the bare minimum of this solution to have the new structured address as instructed in the [EPC guidance 2.1](https://www.europeanpaymentscouncil.eu/document-library/guidance-documents/epc-guidance-document-provision-addresses-under-epc-payment).

### CODE REVIEW

File: account_banking_pain_base/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+from odoo import api, models
+
+logger = logging.getLogger(__name__)
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    def _generate_address_block(self, parent_node, gen_args):
```
Comment (pedrobaeza) [inline account_banking_pain_base/models/res_partner.py]: As you are moving this method, please rename it to something understandable on its own inside `res.partner`, like `_generate_sepa_pain_address_block`.

File: account_banking_pain_base/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                    {"state_name": state_name[0]},
+                    35,
+                    gen_args,
+                )
+
+            ctry = objectify.SubElement(postal_address, "Ctry")
+            ctry.text = self.country_id.code
+
+    @api.model
+    def _improved_street_split(self, street):
```
Comment (pedrobaeza) [inline account_banking_pain_base/models/res_partner.py]: The same for this.

Comment (mihien) [reply account_banking_pain_base/models/res_partner.py]: Good point, I've changed the two names (and one more) to make them more explicit.

**Comment by thomaspaulb** [comment]
@mihien Could I ask for a review on our alternative PR https://github.com/OCA/bank-payment/pull/1550

It's for 14.0 since we are testing it there but can be easily forward ported

It's a bit simpler:

- By default it adds only CtryId and TwnNm (this is the safest bet)
- When hybrid address mode is selected, it adds AdrLine blocks for street and street2 (this should in theory be accepted by banks but it's not accepted in our bank)
- Your PR case could be plugged in on top by adding an extra "sepa_pain09_address_mode = 'structured'" option where you then try to fill in as much details as you can

I'm hoping that with some reviews/small changes our small PR could be merged first, and then this bigger one can still be an improvement on top.

@alexis-via You are probably busy but when you have the time please weigh in on this as well!

**Review by marcos-mendez — CHANGES_REQUESTED** [review]
**Automated Review -- Tests Failed**

### 1. **Root Cause of the Test Failure**
The test fails because the `unidecode` Python library is missing from the test environment, which is required by the `account_banking_pain_base` module. This causes a `DistributionNotFound` error during module loading.

### 2. **Suggested Fix**
Install the missing `unidecode` Python library in the test environment:
```bash
pip install unidecode
```
Alternatively, add `unidecode` to the `external_dependencies` section in the module's `__manifest__.py`:
```python
'external_dependencies': {
    'python': ['unidecode'],
},
```

### 3. **Additional Code Issues**
- **In `account_payment_order.py` line 403–407**: The condition `gen_args["pain_flavor"].startswith("pain.001.001.09")` is hardcoded and may not be future-proof. Consider using a helper method or a more flexible lookup to support additional PAIN versions.
- **In `res_partner.py` line 120**: The `state_name` assignment uses a tuple unpacking:
  ```python
  state_name = (
      self.with_context(
          lang=self.lang or self.env.user.lang
      ).state_id.name,
  )
  ```
  This should be simplified to:
  ```python
  state_name = self.with_context(
      lang=self.lang or self.env.user.lang
  ).state_id.name
  ```

### 4. **Test Improvements**
- **Add tests for `generate_address_block`**: Since this method was removed and replaced, ensure that `test_generate_sepa_pain_address_block` and `test_generate_sepa_pain_unstructured_address_block` cover all PAIN flavors (`pain.001.001.03`, `pain.001.001.09`, `pain.001.003.03`) and edge cases like missing fields.
- **Add tests for `_split_street_into_name_number`**: Include test cases for:
  - Street names with numbers at the end.
  - Street names with numbers at the beginning.
  - Street names with extensions like "bis", "ter", etc.
  - Empty or invalid inputs.
- **Use `SavepointCase`**: For tests that modify data (e.g., `test_generate_sepa_pain_address_block`), use `odoo.tests.common.SavepointCase` to isolate database changes.
- **Tag tests**: Add tags like `@tag('pain')` or `@tag('address')` to improve test organization and filtering.

These changes will improve test coverage and maintainability.

---
⏰ This PR has been open for **87 days**.
💤 Last activity was 43 days ago.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! I found some test failures on this PR and left detailed feedback above. I am happy to discuss or help debug. In the meantime, if any of you get a chance, I would appreciate a look at my open PR(s):

**My open PRs across OCA:**
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward. Thank you!

---
_Environment via [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer): Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

**Comment by victor-champonnois** [comment]
Functional test : I tried to create a payment order for suppliers. But had this error when generating the xml file

```python
RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/home/odoo16/src/odoo/odoo/http.py", line 1846, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/home/odoo16/src/odoo/odoo/service/model.py", line 153, in retrying
    result = func()
  File "/home/odoo16/src/odoo/odoo/http.py", line 1874, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/home/odoo16/src/odoo/odoo/http.py", line 2078, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/home/odoo16/src/odoo/addons/website/models/ir_http.py", line 237, in _dispatch
    response = super()._dispatch(endpoint)
  File "/home/odoo16/src/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/home/odoo16/src/odoo/odoo/http.py", line 761, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/home/odoo16/src/odoo/addons/web/controllers/dataset.py", line 47, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/home/odoo16/src/odoo/addons/web/controllers/dataset.py", line 34, in _call_kw
    return call_kw(Model, method, args, kwargs)
  File "/home/odoo16/src/odoo/odoo/api.py", line 484, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/odoo16/src/odoo/odoo/api.py", line 469, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo16/src/oca/bank-payment/account_payment_order/models/account_payment_order.py", line 430, in open2generated
    payment_file_str, filename = self.generate_payment_file()
  File "/home/odoo16/src/oca/bank-payment/account_banking_sepa_direct_debit/models/account_payment_order.py", line 18, in generate_payment_file
    return super().generate_payment_file()
  File "/home/odoo16/src/oca/bank-payment/account_banking_sepa_credit_transfer/models/account_payment_order.py", line 71, in generate_payment_file
    pain_root = etree.SubElement(xml_root, root_xml_tag)
UnboundLocalError: local variable 'root_xml_tag' referenced before assignment

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    RPCError@https://medor.test.coopiteasy.be/web/assets/354145-c7769b0/web.assets_backend.min.js:999:274
    makeErrorFromResponse@https://medor.test.coopiteasy.be/web/assets/354145-c7769b0/web.assets_backend.min.js:1003:163
    jsonrpc/promise</<@https://medor.test.coopiteasy.be/web/assets/354145-c7769b0/web.assets_backend.min.js:1011:34
```

**Comment by thomaspaulb** [comment]
@mihien I would suggest closing this, since with your project shelved there is not a big chance that you will rebase this on the now-merged simpler solution any time soon?
